### DISCUSSION TITLE
Configurations are now scoped

### DISCUSSION THREAD

**Discussion by chemzqm** [neoclide/coc.nvim#4185] [Show and tell]
A configuration could be one of three different configuration scopes:

- `application` the configuration could only be changed in user configuration
  file.
- `resource` the configuration could be changed in user and workspace folder
  configuration file.
- `language-overridable` the configuration could be changed in user and
  workspace folder configuration file, and can be use used in language scoped
  configuration.

Use language scoped configurations in your coc-settings.json like:
```json
  "[typescript][json]": {
    "diagnostic.enable": true,
    "semanticTokens.enable": true,
    "inlayHint.enable": true,
    "signature.enable": true
  }
```

Deprecated configurations:

- `inlayHint.filetypes`
-  `semanticTokens.filetypes`

They will still work, and will overwrite related scoped configuration.

Since coc.nvim have only supported 3 scoped, some conventions are done for configuration from property scope in package.json:
```js
  if (scope === 'application' || scope === 'window') return ConfigurationScope.WINDOW
  if (scope === 'resource' || scope === 'machine-overridable') return ConfigurationScope.RESOURCE
  if (scope === 'language-overridable') return ConfigurationScope.LANGUAGE_OVERRIDABLE
  return ConfigurationScope.RESOURCE
```

**Comment by chemzqm** [comment]
coc-json need improved to support scoped configurations.

**Reply by chemzqm** [reply]
Fixed on lastest coc.nvim with coc-json 1.7.0

**Comment by neur1n** [comment]
Are users supposed to configure the language scope like this?

```json
  "[c][cpp][python][<other lang>]": {
    "diagnostic.enable": true,
    // ...
  }
```

Is global language scope (say `[*]`) allowed?

**Reply by chemzqm** [reply]
No `[*]`, not needed at all.

**Comment by nullromo** [comment]
It seems to work, however I get this issue:

`Property [c][cpp] is not allowed. (json)`

![`Property [c][cpp] is not allowed. (json)`](https://user-images.githubusercontent.com/8991581/196228213-e6dc6db9-c1df-4c9e-9461-23054716e382.png)

**Reply by nullromo** [reply]
>coc-json need improved to support scoped configurations.

Is this related? I will have a look at `coc-json`.

**Reply by nullromo** [reply]
I posted an issue on coc-json in case other people are searching for this.
